discoverd/server/store.go

```diff
@@ -0,0 +1,1030 @@
+
+	// Broadcast "down" event for instance.
+	if entry.Instance != nil {
+		s.broadcast(&discoverd.Event{
+			Service:  c.Service,
+			Kind:     discoverd.EventKindDown,
+			Instance: entry.Instance,
+		})
+	}
+
+	// Invalidate service leadership.
+	s.invalidateServiceLeader(c.Service, c.Now)
+
+	return nil
+}
+
+// ServiceMeta returns the meta data for a service.
+func (s *Store) ServiceMeta(service string) *discoverd.ServiceMeta {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	return s.serviceMeta(service)
+}
+
+func (s *Store) serviceMeta(service string) *discoverd.ServiceMeta {
+	return s.data.Metas[service]
+}
+
+func (s *Store) SetServiceMeta(service string, meta *discoverd.ServiceMeta) error {
+	// Serialize command.
+	cmd, err := json.Marshal(&setServiceMetaCommand{
+		Service: service,
+		Meta:    meta,
+	})
+	if err != nil {
+		return err
+	}
+
+	index, err := s.raftApply(setServiceMetaCommandType, cmd)
+	if err != nil {
+		return err
+	}
+	meta.Index = index
+
+	return nil
+}
+
+func (s *Store) applySetServiceMetaCommand(cmd []byte, index uint64) error {
+	var c setServiceMetaCommand
+	if err := json.Unmarshal(cmd, &c); err != nil {
+		return err
+	}
+
+	// Verify that the service exists.
+	service := s.data.Services[c.Service]
+	if service == nil {
+		return NotFoundError{Service: c.Service}
+	}
+
+	// If no index is provided then the meta should not be set.
+	curr := s.data.Metas[c.Service]
+	if c.Meta.Index == 0 {
+		if curr != nil {
+			return hh.ObjectExistsErr(fmt.Sprintf("Service metadata for %q already exists, use index=n to set", c.Service))
+		}
+	} else {
+		if curr == nil {
+			return hh.PreconditionFailedErr(fmt.Sprintf("Service metadata for %q does not exist, use index=0 to set", c.Service))
+		} else if curr.Index != c.Meta.Index {
+			return hh.PreconditionFailedErr(fmt.Sprintf("Service metadata for %q exists, but wrong index provided", c.Service))
+		}
+	}
+
+	// Update the meta and set the index.
+	c.Meta.Index = index
+	s.data.Metas[c.Service] = c.Meta
+
+	// Broadcast EventKindServiceMeta event.
+	s.broadcast(&discoverd.Event{
+		Service:     c.Service,
+		Kind:        discoverd.EventKindServiceMeta,
+		ServiceMeta: c.Meta,
+	})
+
+	return nil
+}
+
+// SetServiceLeader manually sets the leader for a service.
+func (s *Store) SetServiceLeader(service, id string) error {
+	// Serialize command.
+	cmd, err := json.Marshal(&setLeaderCommand{
+		Service: service,
+		ID:      id,
+	})
+	if err != nil {
+		return err
+	}
+
+	if _, err := s.raftApply(setLeaderCommandType, cmd); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *Store) applySetLeaderCommand(cmd []byte) error {
+	var c setLeaderCommand
+	if err := json.Unmarshal(cmd, &c); err != nil {
+		return err
+	}
+
+	s.data.Leaders[c.Service] = c.ID
+
+	// Notify new leadership.
+	if entry, ok := s.data.Instances[c.Service][c.ID]; ok && entry.Instance != nil {
+		log.Printf("XXX> SET LEADER: service=%v, inst=%v", c.Service, c.ID)
```

Comment:
This debug print should be removed, but we currently have some debug logging in the broadcast path of the old discoverd that's worth copying into here.